ntoskrnl/se/token.c

```diff
@@ -2220,78 +2244,83 @@ SeTokenCanImpersonate(
     _In_ PTOKEN TokenToImpersonate,
     _In_ SECURITY_IMPERSONATION_LEVEL ImpersonationLevel)
 {
-    BOOLEAN CanImpersonate;
     PAGED_CODE();
 
-    /*
-     * SecurityAnonymous and SecurityIdentification levels do not
-     * allow impersonation.
-     */
-    if (ImpersonationLevel == SecurityAnonymous ||
-        ImpersonationLevel == SecurityIdentification)
-    {
-        return FALSE;
-    }
-
     /* Time to lock our tokens */
     SepAcquireTokenLockShared(ProcessToken);
     SepAcquireTokenLockShared(TokenToImpersonate);
 
-    /* What kind of authentication ID does the token have? */
-    if (RtlEqualLuid(&TokenToImpersonate->AuthenticationId,
-                     &SeAnonymousAuthenticationId))
+    /* Allow impersonation for the process server if it's granted the impersonation privilege */
+    if ((ProcessToken->TokenFlags & TOKEN_HAS_IMPERSONATE_PRIVILEGE) != 0)
```

Comment:
Are you sure this is the correct ordering of checks (verifying this flag, before doing the equalluid for anonymous auth on the TokenToImpersonate)?